heron/common/src/python/argparser.py

```diff
@@ -0,0 +1,196 @@
+          # associated so that we can load the args to a key value
+          # structure
+          args_list = config.insert_bool_values(value.split())
+          args_list_string = ' '.join(args_list)
+          if not  command or not app or not env:
+            Log.warn("heronrc config entry %s does not have key parameters (command:app:env) ",
+                     line)
+            continue
+          if app not in cls.cmdmap:
+            cls.cmdmap[app] = collections.defaultdict(dict)
+
+          if command in cls.cmdmap[app] and env in cls.cmdmap[app][command]:
+            cls.cmdmap[app][command][env] = cls.cmdmap[app][command][env] + ' ' + args_list_string
+          else:
+            cls.cmdmap[app][command][env] = args_list_string
+      Log.debug("RC cmdmap %s", json.dumps(cls.cmdmap))
+    else:
+      Log.warn("%s is not an existing file", effective_rc)
+
+  # for each command / cluster-role-env combination, get the commands from heronrc
+  # remove any duplicates that have already been supplied already  and
+  # present in the command-dictionary
+  @classmethod
+  def get_args_for_command_role(cls, app, command, role):
+    args_for_command_role = ''
+    if app in cls.cmdmap and command in cls.cmdmap[app] and role in cls.cmdmap[app][command]:
+      args_for_command_role = (cls.cmdmap[app][command][role],
+                               args_for_command_role)[cls.cmdmap[app][command][role] is None]
+    return args_for_command_role.split()
+
+  # this is invoked when the parser.parse_args is called
+  #  apply the commands in the following precedence order
+  #  use the defaults in the command line
+
+  def _read_args_from_files(self, arg_strings):
+    new_arg_strings = []
+    command = self.rccommand
+    if len(sys.argv) > 1:
+      command = (sys.argv[1], self.rccommand)[self.rccommand != '']
+    role = self.rcclusterrole
+    if len(sys.argv) > 2:
+      role = (sys.argv[2], self.rcclusterrole)[self.rcclusterrole != '']
+    app = self.prog
+    new_arg_strings.extend(
+        self.get_args_for_command_role(app, command, role))
+    new_arg_strings.extend(
+        self.get_args_for_command_role(app, command, '*'))
+    new_arg_strings.extend(self.get_args_for_command_role(app, '*', '*'))
+    new_arg_strings.extend(self.get_args_for_command_role('*', '*', '*'))
+    arg_strings.extend(new_arg_strings)
+    return arg_strings
+
+  def parse_known_args(self, args=None, namespace=None):
+    namespace, args = super(HeronRCArgumentParser,
+                            self).parse_known_args(args, namespace)
+    dict_ns = namespace.__dict__
+    if self.prog == 'heron':
+      try:
+        for key in dict_ns:
+          val = dict_ns[key]
+          if val is not None and isinstance(val, list) and len(val) > 0:
+            dict_ns[key] = val[0]
+      except Exception:
+        Log.warn("heronrc: unable to clobber arguments (%s,%s ) ", namespace, args)
```

Comment:
this is not a fatal condition. this means a misconfiguration in heronrc file and hence the warning.